### PR TITLE
cryptomb: add queue size statistics

### PR DESCRIPTION
**Pull request by VillePihlava**
Commit Message:

Add queue size statistics to CryptoMb private key provider. The queue size statistics show the amount of simultaneous crypto operations handled at the time of a SIMD (single instruction, multiple data) operation. The statistics consist of a histogram that indicates how often different queue sizes have been processed.

Additional Description:

CryptoMb performance depends on the amount of simultaneous crypto operations processed. Statistics show how often different queue sizes are processed. The bigger the queue size, the better the throughput. The statistics help with debugging performance issues, finding good poll delay values, and show the distribution of different queue sizes.

Risk Level: Low
Testing: Unit testing, manual testing
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: CryptoMb private key provider requires Intel 3rd generation Xeon Scalable server processor for the AVX-512 IFMA instruction set.

### CODE REVIEW

**Review by ggreenway — CHANGES_REQUESTED** [review]
This would make more sense as a histogram; you're basically re-implementing a histogram in this code. cc @jmarantz

File: contrib/cryptomb/private_key_providers/source/cryptomb_stats.cc [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace Envoy {
+namespace Extensions {
+namespace PrivateKeyMethodProvider {
+namespace CryptoMb {
+
+CryptoMbStats::CryptoMbStats(Stats::Scope& scope, uint32_t max_queue_size,
+                             absl::string_view stats_prefix,
+                             absl::string_view queue_size_stat_prefix)
+    : stat_name_pool_(scope.symbolTable()), stats_prefix_(stat_name_pool_.add(stats_prefix)) {
```
Comment (jmarantz) [inline contrib/cryptomb/private_key_providers/source/cryptomb_stats.cc:15]: stats_prefix_ is not referenced after the constructor so you can drop the member variable and make this a local.

File: contrib/cryptomb/private_key_providers/test/ops_test.cc [hunk 2]
```diff
@@ -430,6 +458,105 @@ TEST(CryptoMbProviderTest, TestRSATimer) {
...
+  RSA_get0_key(rsa, &n, &e, &d);
+  fakeIpp->setRsaKey(n, e, d);
+  CryptoMbStats stats(server_stats_store, CryptoMbQueue::MULTIBUFF_BATCH, kStatsPrefix,
+                      kQueueSizeStatPrefix);
+
+  CryptoMbQueue queue(std::chrono::milliseconds(200), KeyType::Rsa, 1024, fakeIpp, *dispatcher,
+                      stats);
+
+  size_t in_len = 32;
+  uint8_t in[32] = {0x7f};
```
Comment (jmarantz) [inline contrib/cryptomb/private_key_providers/test/ops_test.cc:479]: there's a lot of repeated blocks of code in this test file. WDYT of factoring some of those out into methods/members of a test class for CryptoMbProviderTest, and shortening all these tests to hopefully make them more readable?

Then we can document some of the code that's mysterious like what this 0x7f is supposed to represent, or why `in` is 32 bytes long.

**Review by jmarantz — COMMENTED** [review]
looks great generally, with one concern/question about how frequently we are going to be doing stats symbol-table lookups, and a few style/readability nits.

**Comment by jmarantz** [comment]
RE @ggreenway 's comment about histograms: I agree that would be a lot better, and that changes a few of my suggestions also. One thing to note is that the XDR histograms auto-decide buckets and I'm not sure if you need something more precise for this. I'm guessing auto-bucketed histograms would be fine.

**Comment by VillePihlava** [comment]
Thanks for the comments!

@jmarantz @ggreenway I tried a histogram for this at one point, but the mentioned auto-decided buckets were confusing. Keeping each queue size as a separate counter shows the exact kind of queue sizes that come up. For example, the information on how frequently the biggest queue size comes up can be especially useful. However, I do agree that the histogram would be a better solution if it worked this way.

**Comment by ggreenway** [comment]
There are two things happening with histograms: the internal storage auto-decides buckets (but is probably quite accurate). And the display from `/stats` displays a configured (not auto-determined) set of buckets. There is a default set of buckets configured, but that can be specified in the bootstrap config under `stats_config`. Also, some stat sinks (such as statsd) don't use either of those mechanisms, and send the raw data.

Can you clarify what the data looked like when you tried with a histogram? It may have just been that you need to configure the displayed buckets to get the results you need.

/wait-any

File: contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc [hunk 3]
```diff
@@ -330,10 +330,10 @@ ssl_private_key_result_t rsaPrivateKeyDecryptForTest(CryptoMbPrivateKeyConnectio
 }
 
 CryptoMbQueue::CryptoMbQueue(std::chrono::milliseconds poll_delay, enum KeyType type, int keysize,
-                             IppCryptoSharedPtr ipp, Event::Dispatcher& d)
+                             IppCryptoSharedPtr ipp, Event::Dispatcher& d, CryptoMbStats& stats)
     : us_(std::chrono::duration_cast<std::chrono::microseconds>(poll_delay)), type_(type),
-      key_size_(keysize), ipp_(ipp),
-      timer_(d.createTimer([this]() -> void { processRequests(); })) {
+      key_size_(keysize), ipp_(ipp), timer_(d.createTimer([this]() -> void { processRequests(); })),
```
[the remark this answers is not in the document]
Comment (VillePihlava) [reply contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc]: Removed `-> void` from two places.

File: contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc [hunk 4]
```diff
@@ -359,6 +359,8 @@ void CryptoMbQueue::addAndProcessEightRequests(CryptoMbContextSharedPtr mb_ctx)
 
 void CryptoMbQueue::processRequests() {
   if (type_ == KeyType::Rsa) {
+    // Increment correct queue size statistic.
+    stats_.getQueueSizeCounters()[request_queue_.size() - 1].get().inc();
```
[the remark this answers is not in the document]
Comment (VillePihlava) [reply contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc]: Removed after changing counters to histogram.

File: contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc [hunk 5]
```diff
@@ -486,7 +488,9 @@ CryptoMbPrivateKeyMethodProvider::CryptoMbPrivateKeyMethodProvider(
         CryptoMbPrivateKeyMethodConfig& conf,
     Server::Configuration::TransportSocketFactoryContext& factory_context, IppCryptoSharedPtr ipp)
     : api_(factory_context.api()),
-      tls_(ThreadLocal::TypedSlot<ThreadLocalData>::makeUnique(factory_context.threadLocal())) {
+      tls_(ThreadLocal::TypedSlot<ThreadLocalData>::makeUnique(factory_context.threadLocal())),
+      stats_(factory_context.scope(), CryptoMbQueue::MULTIBUFF_BATCH, "cryptomb",
+             "rsa_queue_size_") {
```
[the remark this answers is not in the document]
Comment (VillePihlava) [reply contrib/cryptomb/private_key_providers/source/cryptomb_private_key_provider.cc]: This happens at config-time, I added a comment about it.

File: contrib/cryptomb/private_key_providers/test/ops_test.cc [hunk 6]
```diff
@@ -382,8 +407,11 @@ TEST(CryptoMbProviderTest, TestRSATimer) {
   const BIGNUM *e, *n, *d;
   RSA_get0_key(rsa, &n, &e, &d);
   fakeIpp->setRsaKey(n, e, d);
```
[the remark this answers is not in the document]
Comment (VillePihlava) [reply contrib/cryptomb/private_key_providers/test/ops_test.cc]: Refactored tests in ops_test.cc and changed this.

File: contrib/cryptomb/private_key_providers/source/cryptomb_stats.cc (hunk 1 above)
Comment (VillePihlava) [reply contrib/cryptomb/private_key_providers/source/cryptomb_stats.cc]: This does not exist anymore after the histogram change.

File: contrib/cryptomb/private_key_providers/test/ops_test.cc (hunk 2 above)
Comment (VillePihlava) [reply contrib/cryptomb/private_key_providers/test/ops_test.cc]: Refactored tests in ops_test.cc and added comments.

**Comment by VillePihlava** [comment]
@ggreenway  Thank you for the explanation! I changed the counters back to a histogram and the statsd solution worked as you explained.

My initial confusion came from the way the histogram stats are displayed from the `/stats` endpoint, something like this:

`listener.0.0.0.0_9000.cryptomb.rsa_queue_sizes: P0(nan,1.0) P25(nan,8.024987213802275) P50(nan,8.049991475868183) P75(nan,8.074995737934092) P90(nan,8.089998295173636) P95(nan,8.094999147586819) P99(nan,8.098999829517364) P99.5(nan,8.099499914758681) P99.9(nan,8.099899982951737) P100(nan,8.1)`

The `/stats/prometheus` endpoint is a bit more helpful for seeing individual queue sizes, but the buckets only have an upper bound and include the queue sizes below it:

```
envoy_listener_cryptomb_rsa_queue_sizes_bucket{envoy_listener_address="0.0.0.0_9000",le="2"} 1
envoy_listener_cryptomb_rsa_queue_sizes_bucket{envoy_listener_address="0.0.0.0_9000",le="3"} 1
envoy_listener_cryptomb_rsa_queue_sizes_bucket{envoy_listener_address="0.0.0.0_9000",le="4"} 2
envoy_listener_cryptomb_rsa_queue_sizes_bucket{envoy_listener_address="0.0.0.0_9000",le="5"} 3
envoy_listener_cryptomb_rsa_queue_sizes_bucket{envoy_listener_address="0.0.0.0_9000",le="6"} 4
envoy_listener_cryptomb_rsa_queue_sizes_bucket{envoy_listener_address="0.0.0.0_9000",le="7"} 9
envoy_listener_cryptomb_rsa_queue_sizes_bucket{envoy_listener_address="0.0.0.0_9000",le="8"} 10
envoy_listener_cryptomb_rsa_queue_sizes_bucket{envoy_listener_address="0.0.0.0_9000",le="9"} 58667
```

When this is in the config:
```
stats_config:
  histogram_bucket_settings:
    match:
      contains: "cryptomb"
    buckets: [2,3,4,5,6,7,8,9]
```

Would there be an easy way to see individual queue sizes for this histogram, for example, from the `/stats` endpoint?

**Comment by ggreenway** [comment]
> Would there be an easy way to see individual queue sizes for this histogram, for example, from the `/stats` endpoint?

I don't think there is currently. The base `/stats` endpoint just prints the summary that you see. That could be an interesting feature though, via something like `/stats?histogram_buckets=1` or something and print out the bucket counts. It wouldn't be hard to implement (see how it's done in the prometheus exporter; would probably be a handful of lines of code).

File: contrib/cryptomb/private_key_providers/test/ops_test.cc [hunk 7]
```diff
@@ -38,398 +33,343 @@ ssl_private_key_result_t rsaPrivateKeyDecryptForTest(CryptoMbPrivateKeyConnectio
...
+  Event::SimulatedTimeSystem time_system_;
+  Event::DispatcherPtr dispatcher_;
+  std::shared_ptr<FakeIppCryptoImpl> fakeIpp_;
+  CryptoMbStats stats_;
+
+  // Result of an operation.
+  ssl_private_key_result_t res_;
+
+  // A size for signing and decryption operation input chosen for tests.
+  static constexpr size_t IN_LEN = 32;
```
Comment (jmarantz) [inline contrib/cryptomb/private_key_providers/test/ops_test.cc:92]: Const member variables follow member-variable naming convention, not macro convention, so call these in_len_, in_, etc.

File: contrib/cryptomb/private_key_providers/test/ops_test.cc [hunk 8]
```diff
@@ -38,398 +33,343 @@ ssl_private_key_result_t rsaPrivateKeyDecryptForTest(CryptoMbPrivateKeyConnectio
...
+  static constexpr uint8_t IN[IN_LEN] = {0x7f};
+
+  // Maximum size of out_ in all tests cases.
+  static constexpr size_t MAX_OUT_LEN = 128;
+  uint8_t out_[MAX_OUT_LEN] = {0};
+
+  // Size of output in out_ from an operation.
+  size_t out_len_ = 0;
+
+  const std::string QUEUE_SIZE_HISTOGRAM_NAME = "cryptomb.rsa_queue_sizes";
```
Comment (jmarantz) [inline contrib/cryptomb/private_key_providers/test/ops_test.cc:103]: ditto.

**Review by jmarantz — COMMENTED** [review]
looks great; just a few nits. Will you follow-up with another PR to add detail to the /stats endpoint for histograms?

File: contrib/cryptomb/private_key_providers/test/ops_test.cc (hunk 7 above)
Comment (VillePihlava) [reply contrib/cryptomb/private_key_providers/test/ops_test.cc]: Fixed.

File: contrib/cryptomb/private_key_providers/test/ops_test.cc (hunk 8 above)
Comment (VillePihlava) [reply contrib/cryptomb/private_key_providers/test/ops_test.cc]: Fixed.

**Comment by VillePihlava** [comment]
I can start looking into a follow-up PR on the `/stats` endpoint.

**Review by jmarantz — APPROVED** [review]
Can you add an issue for the /stats histogram detail follow-up? I'll assign it to you.

I'll pass over to Greg for a senior maintainer pass.

**Review by ggreenway — APPROVED** [review]
LGTM

**Comment by VillePihlava** [comment]
The linked pull request https://github.com/envoyproxy/envoy/pull/19586 is the follow-up to the discussion in this pull request.

**Pull request merged**